### ISSUE TITLE
panic: zig compiler bug: NeededSourceLocation

### ISSUE THREAD

**Issue by aquamorta** [ziglang/zig#17975]
### Zig Version

0.11.0

### Steps to Reproduce and Observed Behavior

```zig
This small program leads to a compiler panic:

const std = @import("std");
// works with normal` chars crashes with unicode char
const LINE = .{'—'} ** 10;

pub fn main() !void {
    std.debug.print("line:{s}", .{LINE ++ "\n"});
}


```
### Expected Behavior

It either should compile or the compiler should emit an error message.

**Comment by aquamorta** [comment]
Actually I think it should report an error here, because LINE is an array of comp_time integers and not an []u8.

**Comment by Vexu** [comment]
Stack trace:
```
/home/vexu/Documents/zig/zig/src/Sema.zig:2443:27: 0x9dc4200 in errMsg__anon_115120 (zig)
    if (src == .unneeded) return error.NeededSourceLocation;
                          ^
/home/vexu/Documents/zig/zig/src/Sema.zig:2455:21: 0x983ce50 in fail__anon_97908 (zig)
    const err_msg = try sema.errMsg(block, src, format, args);
                    ^
/home/vexu/Documents/zig/zig/src/Sema.zig:28239:25: 0x93f0e2e in coerceExtra (zig)
                        return sema.fail(block, inst_src, "type '{}' cannot represent integer value '{}'", .{ dest_ty.fmt(mod), val.fmtValue(inst_ty, mod) });
                        ^
/home/vexu/Documents/zig/zig/src/Sema.zig:27799:21: 0x91375c4 in coerce (zig)
        else => |e| return e,
                    ^
/home/vexu/Documents/zig/zig/src/Sema.zig:13795:47: 0x97260b9 in zirArrayCat (zig)
                const coerced_elem_val_inst = try sema.coerce(block, resolved_elem_ty, elem_val_inst, .unneeded);
                                              ^
/home/vexu/Documents/zig/zig/src/Sema.zig:1011:46: 0x93c66a3 in analyzeBodyInner (zig)
            .array_cat                    => try sema.zirArrayCat(block, inst),
                                             ^
```

**Comment by nektro** [comment]
> Actually I think it should report an error here, because LINE is an array of comp_time integers and not an []u8.

note that that wouldn't intrinsically be an issue if the character fit within ascii/a `u8`. but `'—'` is `2014` as an int.

conversely, `const LINE = "—".* ** 10;` would likely give you your expected outcome

**Comment by schmee** [comment]
Seems fixed as of `0.12.0-dev.2760+6fe90a913`:

```
main.zig:6:35: error: type 'u8' cannot represent integer value '8212'
    std.debug.print("line:{s}", .{LINE ++ "\n"});
                                  ^~~~
referenced by:
    callMain: /Users/user/repos/zig/lib/std/start.zig:511:32
    callMainWithArgs: /Users/user/repos/zig/lib/std/start.zig:469:12
    remaining reference traces hidden; use '-freference-trace' to see all reference traces
```

**Comment by Vexu** [comment]
Fixed by #18162
